tools/client/http.go

```diff
@@ -0,0 +1,145 @@
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	httpContentType = "application/json"
+)
+
+var (
+	successCodes = []int{
+		http.StatusOK,
+		http.StatusCreated,
+	}
+)
+
+type httpClient struct {
+	u  *url.URL
+	cl http.Client
+}
+
+func newHTTPClient(raw string) (*httpClient, error) {
+	u, err := url.Parse(raw)
+	if err != nil {
+		return nil, errors.Wrap(err, "invalid url")
+	}
+
+	return &httpClient{u: u}, nil
+}
+
+func (c *httpClient) url(p ...string) string {
+	b := *c.u
+	b.Path = filepath.Join(b.Path, filepath.Join(p...))
+
+	return b.String()
+}
+
+func (c *httpClient) process(response *http.Response, output interface{}, expect ...int) error {
+	defer response.Body.Close()
+
+	if len(expect) == 0 {
+		expect = successCodes
+	}
+
+	in := func(i int, l []int) bool {
+		for _, x := range l {
+			if x == i {
+				return true
+			}
+		}
+		return false
+	}
+
+	dec := json.NewDecoder(response.Body)
+	if !in(response.StatusCode, expect) {
+		var output struct {
+			E string `json:"error"`
+		}
+
+		if err := dec.Decode(&output); err != nil {
+			return errors.Wrapf(err, "failed to load error while processing invalid return code of: %s", response.StatusCode)
+		}
+
+		return fmt.Errorf("%s: %s", response.Status, output.E)
+	}
+
+	if output == nil {
+		//discard output
+		return nil
```

Comment:
```suggestion
		output = ioutil.Discard
```
So we are sure to consume eveything from the body in any case and allow to reuse the connection ?

Reply:
Idd, but this thin should be ioutil.ReadAll(response.Body)

Reply:
well you already consume the body in the `dec.Decode` underneath, but this you are right better going it like you propose